catalog/transformer/catalog-transformer-service-atom/src/main/java/ddf/catalog/transformer/response/query/atom/AtomTransformer.java

```diff
@@ -473,4 +484,28 @@ private Link addLinkHelper(Action action, Entry entry, String linkType, String m
     }
     return georssPositions;
   }
+
+  private int getCount(SourceResponse queryResponse) {
```

Comment:
✏️  `queryResponse` should be annotated with`@Nullable`

Reply:
Do we put `@Nullable` on private methods?

Reply:
If we can help another developer down the road by letting them know a response could be null, I'd say we should.

Reply:
This method returns a simple data type.  It's never null.

Reply:
No, `queryResponse` can be null. Arguments can be marked with `@Nullable` as well.

Reply:
I'm aware of that. My point was in reference to "If we can help another developer down the road by letting them know a response could be null, I'd say we should."

Reply:
I think he meant parameter in this case although the comment applies to both return value and parameters, we should always annotate them with @Nullable even private ones.

Reply:
I'm curious why `we should` be annotating this; is there some guideline that can be referenced?  The fact that the method is annotated with `@nullable` is seemingly of no consequence to other developers (and the null check in the preceding line makes it obvious).  Assuming we are discussing `javax.annotation.Nullable`, this is from the documentation:

`This annotation is useful mostly for overriding a {@link Nonnull} annotation. Static analysis tools should generally treat the annotated items as though they had no annotation, unless they are configured to minimize false negatives. Use {@link CheckForNull} to indicate that the element value should always be checked for a null value.`

Reply:
As I was told by @lessarderic, all methods that returns `null` and all parameters that accepts `null` should now be annotated with @Nullable. This is the convention that was adopted not too long ago. By default, methods do not return `null` and do not accept `null`. So even though looking at the method might tell you, we still want to annotate them to make it clear that this was the intent. Think of it as proper documentation.

Reply:
Where is this convention documented?  This seems to be at odds with the documentation linked above for `@nullable`

Reply:
I don't see documentation of conventions related to handling `null` for DDF or other Connexta projects.
In this code base and others, the pattern I've noticed is to never pass `null` into something unless it is clear that `null` is an expected value. A method's signature should be sufficient to explain what the method does and how to use it. The `@Nullable` annotation can clarify that `null` as a valid possible value without having to look at the body of the method.
The way that I see it, without `@Nullable`, this method looks like it expects a `SourceResponse` argument. Much of DDF's code seems to assume that arguments are not `null` unless they are marked with `@Nullable`. (See `WorkspaceMetacardImpl` as one such example that I happened upon while working.) Adding `@Nullable` indicates that the first argument can actually be either a `SourceResponse` or `null`.
According to the `Nullable` documentation that you posted, we should be using `@Nonnull` all over the place; however, most things in code are not intended to be null, so using `@Nonnull` on every `Object` expected to not be `null` would result in a high amount of code clutter. Using `@Nullable` when `null` is expected is a less safe, but less cluttered and more common alternative.

Reply:
This is exactly why they switch the convention in DDF from using the @Nonnull annotation to only @Nullable. We are slowly updating code that still use the old @Nonnull annotation as we get into those files.

Reply:
My point is that if this is a DDF convention it should be documented somewhere on the Codice Confluence page if it isn't already.  Right now it seems to be a matter of preference and now a point of contention.  As @paouelle has stated, the drive to use `@Nullable` is recent, and I would argue that this annotation is absent more than it is present in this large code base.  Do we intend on going through the entire code base, adding `@Nullable` where applicable?

Reply:
Find DDF guidelines at https://codice.atlassian.net/wiki/spaces/DDF/pages/6291514/Code+Comment+Guidelines which described this convention

Reply:
Unless i'm mistaken, this describes a convention for annotations for `all externally-facing public and protected methods` and not for private methods.

Reply:
Should we really stop there? At the end of the day, it is meant to clarify the code and the intent.

Reply:
I would personally prefer private methods including it as well -- that way if I have to go and update someone elses code somewhere the intents and capability of a method and what should be passed to it are communicated clearly from the beginning.  Having it only on some methods and not others (public / protected and not default / private) seems to add ambiguity to a methods contract.

Also we can use this interface for static analysis benefit, though if we only do it on some of the methods and not others then the benefit of the static analysis is lessened (or wrong even).

Reply:
I'm with @dcruver. I see zero value in adding that annotation to private methods. Traceability of private methods is trivial for a developer to ensure they don’t break anything, plus static analysis will pick up any potential null accesses missed by humans.

Code clutter is a real concern; Java is overly verbose by nature and adding extra noise for no return on investment makes it more onerous and less desirable as a development language.

Reply:
"to a methods contract": I think you're confusing the scope of the term "contract".  By definition, private methods aren't part of the interface or contract. The intent of @notnull and @nullable is to help clarify the contract of a method that's intentionally being designed for extension. I don't think that blindly adding these annotations to every method parameter and return value is either practical or helpful.

Reply:
Regardless of any opinions, they are just 'guidelines' in the end. I don't think anyone would be required to do anything they don't want to. Probably akin to adding javadoc to a private method: Nothing says you have to but nothing also says you can't. 😄

Reply:
So those of us that want to be gross and verbose can do as we wish, and those that don't wish to don't have to 😄

Reply:
While this is a style suggestion, I think one of the goals here was to establish some precedent so that it isn't a continuous debate.

Reply:
I think in a org of 50 developers working in the same codebase, the fact we've gotten this far harmoniously is a miracle. I don't think a little wiggle room will kill anyone. 

Since it is a guideline, likening it to javadoc on a private method is adequate. It is by no means required but also not disallowed completely. 

The guidelines also state:
```Package private (default visibility), private methods, or otherwise internal code do not have to include JavaDoc comments. If their purpose cannot be clearly conveyed by their name and the name of their parameters, use just as much javadoc as necessary to communicate the behavior.```

Reply:
With this guideline in mine, how does a parameter name say wether it accepts null or not if it ain't annotated and if one doesn't want to start looking at the code to figure out if it should be or if it was meant to be?

Reply:
@paouelle I think the point is feel free to use it yourself, but as of now others are not required to use it.

Reply:
I think it would also be situational; in this case, it's obvious that it can be null.